wrangler-storage/src/main/java/io/cdap/wrangler/store/recipe/RecipeStore.java

```diff
@@ -0,0 +1,168 @@
+    TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(TABLE_ID);
+
+      String recipeName = recipeRow.getRecipe().getRecipeName();
+      RecipeRow oldRecipeWithName = getRecipeByName(table, recipeName, recipeId.getNamespace(), failIfNotFound);
+      if (!failIfNotFound && oldRecipeWithName != null) {
+        throw new RecipeAlreadyExistsException(String.format("recipe with name '%s' already exists", recipeName));
+      }
+
+      RecipeRow oldRecipe = getRecipeInternal(table, recipeId, failIfNotFound);
+      RecipeRow newRecipe = recipeRow;
+      if (oldRecipe != null) {
+        Recipe updatedRecipe = Recipe.builder(newRecipe.getRecipe())
+          .setCreatedTimeMillis(oldRecipe.getRecipe().getCreatedTimeMillis()).build();
+        newRecipe = RecipeRow.builder(recipeId.getNamespace(), updatedRecipe).build();
+      }
+
+      Collection<Field<?>> fields = getRecipeKeys(recipeId);
+      fields.add(Fields.stringField(RECIPE_NAME_FIELD, newRecipe.getRecipe().getRecipeName()));
+      fields.add(Fields.longField(CREATE_TIME_COL, newRecipe.getRecipe().getCreatedTimeMillis()));
+      fields.add(Fields.longField(UPDATE_TIME_COL, newRecipe.getRecipe().getUpdatedTimeMillis()));
+      fields.add(Fields.stringField(RECIPE_INFO_COL, GSON.toJson(newRecipe)));
+
+      table.upsert(fields);
+    });
+  }
+
+  private RecipeRow getRecipeInternal(StructuredTable table, RecipeId recipeId, boolean failIfNotFound)
+    throws RecipeNotFoundException, IOException {
+    Optional<StructuredRow> row = table.read(getRecipeKeys(recipeId));
+    if (!row.isPresent()) {
+      if (!failIfNotFound) {
+        return null;
+      }
+      throw new RecipeNotFoundException(String.format("recipe with id '%s' does not exist", recipeId.getRecipeId()));
+    }
+    return GSON.fromJson(row.get().getString(RECIPE_INFO_COL), RecipeRow.class);
+  }
+
+  private RecipeRow getRecipeByName(StructuredTable table, String recipeName,
+                                    NamespaceSummary summary, boolean failIfNotFound) throws IOException {
+    Collection<Field<?>> keys = getNamespaceKeys(summary);
+    keys.add(Fields.stringField(RECIPE_NAME_FIELD, recipeName));
+    Range range = Range.singleton(keys);
+    CloseableIterator<StructuredRow> iterator = table.scan(range, 1);
+    if (!iterator.hasNext()) {
+      if (!failIfNotFound) {
+        return null;
+      }
+      throw new RecipeNotFoundException(String.format("recipe with name '%s' does not exist", recipeName));
+    }
+    return GSON.fromJson(iterator.next().getString(RECIPE_INFO_COL), RecipeRow.class);
+  }
+
+  private Collection<Field<?>> getRecipeKeys(RecipeId recipeId) {
+    List<Field<?>> keys = new ArrayList<>(getNamespaceKeys(recipeId.getNamespace()));
+    keys.add(Fields.stringField(RECIPE_ID_FIELD, recipeId.getRecipeId()));
+    return keys;
+  }
+
+  private Collection<Field<?>> getNamespaceKeys(NamespaceSummary namespace) {
```

Comment:
I see the same function in `WorkspaceStore.java`, can we reuse the same function or may be move the function to a common class, so both Stores can use the same function?